### ISSUE TITLE
Java 8 Lambda variables don't need to be final

### ISSUE THREAD

**Issue by dharrigan** [checkstyle/checkstyle#747]
Hi,

Checkstyle 6.4.1
Java 8u40

Running Checkstyle 6.4.1 against some Java 8 code, produces a lot of these type of warnings (a small example is shown below):

```
[ant:checkstyle] 
Orders.java:150:26: warning: Variable 't' should be declared final.
[ant:checkstyle] 
Orders.java:150:29: warning: Variable 'u' should be declared final.
```

An example of the code that is valid in Java 8 (and is being checked by Checkstyle) is shown below. Putting "final" before the variables (t and u) produces syntax errors, therefore Checkstyle should not flag these conditions as needing final keywords:

```
private void addTotalValueOfOrder(final AugmentedOrder order) {
    final BigDecimal totalValueOfOrder = order.getTransactions().stream()
            .reduce(BigDecimal.ZERO,
                    (t, u) -> t.add(u.getAmount()),
                    BigDecimal::add);
    order.setTotalValueOfOrder(totalValueOfOrder.toPlainString());
 }
```

Thank you.

-=david=-

**Comment by romani** [comment]
Please provide config, to be completely sure that we speak about the your problem

**Comment by dharrigan** [comment]
Hi,

Here you go. If you need anything else, please do let me know :-)

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
        <property name="tabWidth" value="4"/>
        <module name="FileContentsHolder"/>
        <module name="ConstantName"/>
        <module name="LocalFinalVariableName"/>
        <module name="LocalVariableName"/>
        <module name="MemberName"/>
        <module name="MethodName"/>
        <module name="PackageName"/>
        <module name="ParameterName"/>
        <module name="StaticVariableName"/>
        <module name="TypeName"/>
        <module name="AvoidStarImport"/>
        <module name="IllegalImport"/>
        <module name="RedundantImport"/>
        <module name="LineLength">
            <property name="max" value="256"/>
            <property name="tabWidth" value="4"/>
        </module>
        <module name="MethodLength"/>
        <module name="ParameterNumber"/>
        <module name="EmptyForIteratorPad"/>
        <module name="MethodParamPad"/>
        <module name="NoWhitespaceAfter">
            <property name="tokens" value="BNOT,DEC,DOT,INC,LNOT,UNARY_MINUS,UNARY_PLUS"/>
        </module>
        <module name="NoWhitespaceBefore"/>
        <module name="OperatorWrap"/>
        <module name="ParenPad"/>
        <module name="TypecastParenPad">
            <property name="tokens" value="RPAREN"/>
        </module>
        <module name="WhitespaceAfter">
            <property name="tokens" value="COMMA,SEMI"/>
        </module>
        <module name="WhitespaceAround">
            <property name="tokens" value="ASSIGN,BAND,BAND_ASSIGN,BOR,BOR_ASSIGN,BSR,BSR_ASSIGN,BXOR,BXOR_ASSIGN,COLON,DIV,DIV_ASSIGN,EQUAL,GE,GT,LAND,LE,LITERAL_ASSERT,LITERAL_CATCH,LITERAL_DO,LITERAL_ELSE,LITERAL_FINALLY,LITERAL_FOR,LITERAL_IF,LITERAL_RETURN,LITERAL_SYNCHRONIZED,LITERAL_TRY,LITERAL_WHILE,LOR,LT,MINUS,MINUS_ASSIGN,MOD,MOD_ASSIGN,NOT_EQUAL,PLUS,PLUS_ASSIGN,QUESTION,SL,SLIST,SL_ASSIGN,SR,SR_ASSIGN,STAR,STAR_ASSIGN,LITERAL_ASSERT,TYPE_EXTENSION_AND"/>
        </module>
        <module name="ModifierOrder"/>
        <module name="RedundantModifier"/>
        <module name="AvoidNestedBlocks">
            <property name="allowInSwitchCase" value="true"/>
        </module>
        <module name="EmptyBlock"/>
        <module name="LeftCurly"/>
        <module name="NeedBraces"/>
        <module name="RightCurly"/>
        <module name="EmptyStatement"/>
        <module name="EqualsHashCode"/>
        <module name="IllegalInstantiation">
            <property name="classes" value="java.lang.RuntimeException"/>
        </module>
        <module name="InnerAssignment"/>
        <module name="MagicNumber"/>
        <module name="MissingSwitchDefault"/>
        <module name="SimplifyBooleanExpression"/>
        <module name="SimplifyBooleanReturn"/>
        <module name="FinalClass"/>
        <module name="InterfaceIsType"/>
        <module name="VisibilityModifier"/>
        <module name="ArrayTypeStyle"/>
        <module name="TodoComment"/>
        <module name="UpperEll"/>
        <module name="UnusedImports"/>
        <module name="AbstractClassName"/>
        <module name="AnnotationUseStyle"/>
        <module name="MissingOverride"/>
        <module name="SuppressWarnings"/>
        <module name="GenericWhitespace"/>
        <module name="DefaultComesLast"/>
        <module name="DeclarationOrder"/>
        <module name="EqualsAvoidNull"/>
        <module name="FallThrough"/>
        <module name="FinalLocalVariable">
            <property name="tokens" value="PARAMETER_DEF,VARIABLE_DEF"/>
        </module>
        <module name="ModifiedControlVariable"/>
        <module name="MultipleStringLiterals">
            <property name="allowedDuplicates" value="3"/>
        </module>
        <module name="MultipleVariableDeclarations"/>
        <module name="NoFinalizer"/>
        <module name="PackageDeclaration"/>
        <module name="ParameterAssignment"/>
        <module name="ReturnCount">
            <property name="max" value="5"/>
        </module>
        <module name="StringLiteralEquality"/>
        <module name="ArrayTrailingComma"/>
        <module name="BooleanExpressionComplexity"/>
        <module name="ClassDataAbstractionCoupling">
            <property name="max" value="30"/>
        </module>
        <module name="ClassFanOutComplexity">
            <property name="max" value="50"/>
        </module>
        <module name="CyclomaticComplexity">
            <property name="max" value="20"/>
        </module>
        <module name="JavaNCSS">
            <property name="methodMaximum" value="75"/>
        </module>
        <module name="NPathComplexity"/>
        <module name="Indentation"/>
        <module name="MissingDeprecated"/>
        <module name="ClassTypeParameterName"/>
        <module name="MethodTypeParameterName"/>
        <module name="AnonInnerLength">
            <property name="max" value="75"/>
        </module>
        <module name="OuterTypeNumber"/>
        <module name="ExecutableStatementCount">
            <property name="max" value="60"/>
            <property name="tokens" value="INSTANCE_INIT,STATIC_INIT,METHOD_DEF,CTOR_DEF"/>
        </module>
        <module name="ExplicitInitialization"/>
        <module name="IllegalCatch"/>
        <module name="IllegalThrows"/>
        <module name="IllegalType">
            <property name="tokens" value="METHOD_DEF,PARAMETER_DEF,VARIABLE_DEF"/>
        </module>
        <module name="NestedIfDepth">
            <property name="max" value="2"/>
        </module>
        <module name="NestedTryDepth">
            <property name="max" value="2"/>
        </module>
        <module name="SuperClone"/>
        <module name="SuperFinalize"/>
        <module name="UnnecessaryParentheses"/>
        <module name="MutableException"/>
        <module name="ThrowsCount">
            <property name="max" value="5"/>
        </module>
        <module name="UncommentedMain"/>
        <module name="RegexpSinglelineJava">
            <property name="format" value="System\.out\.println"/>
            <property name="message" value="Use of System.out.println is prohibited!"/>
            <property name="ignoreCase" value="true"/>
        </module>
        <module name="Indentation"/>
        <module name="FinalParameters">
            <property name="tokens" value="METHOD_DEF,CTOR_DEF,LITERAL_CATCH"/>
        </module>
        <module name="TrailingComment"/>
        <module name="JavadocType"/>
    </module>
    <module name="Translation"/>
    <module name="FileLength">
        <property name="max" value="1024"/>
    </module>
    <module name="FileTabCharacter">
        <property name="eachLine" value="true"/>
    </module>
    <module name="SuppressionFilter">
        <property name="file" value="config/checkstyle/suppressions.xml"/>
    </module>
    <module name="SuppressionCommentFilter"/>
</module>
```

**Comment by romani** [comment]
fixed, will be in 6.5

**Comment by dharrigan** [comment]
w00t!

Thanks.

-=david=-
